proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,521 @@
+4. A local computed property representing  `wrappedValue` will be synthesized by the compiler and named per the original – non-prefixed – parameter name. Moreover, the accessors will mirror the `wrappedValue` accessors. Lastly, a setter will _only_ be synthesized for the local property if the `wrappedValue`'s setter is `nonmutating`, or if the wrapper is a reference type.
+5. If the property wrapper defines a `projectedValue` property, a local computed property representing `projectedValue` will be synthesized by the compiler and named per the original parameter name prefixed with a dollar sign (`$`). The same accessor rules for `wrappedValue` apply to `projectedValue`.
+6. When passing an argument to a property wrapper parameter, the compiler will wrap the argument in the appropriate `init(wrappedValue:)` call.
+
+#### Transformation example:
+
+```swift
+@propertyWrapper
+struct Lowercased {
+
+  init(wrappedValue: String) { ... }
+    
+    
+  var wrappedValue: String {
+    get { ... }
+    set { ... }
+  }
+  
+}
+
+func postUrl(
+  @Lowercased urlString: String
+) { ... }
+
+
+postUrl(urlString: "mySite.xyz/myUnformattedUsErNAme")
+```
+
+In the above code, the `postUrl(urlString:)` function and its caller are equivalent to:
+
+```swift
+func postUrl(urlString _urlString: Lowercased) {
+
+  var urlString: String {
+    get { _ urlString.wrappedValue }
+    // The setter accessor is not synthesized
+    // because the setter of the 'wrappedValue' 
+    // property of 'Lowercased' is mutating.
+  }
+
+
+  ...
+  
+}
+
+
+postUrl(urlString: Lowercased(wrappedValue: "mySite.xyz/myUnformattedUsErNAme"))
+```
+
+#### Restrictions on property-wrapper function-parameters
+
+##### `@autoclosure`
+
+Function parameters with a property-wrapper custom-attribute cannot have an `@autoclosure` type, because `@autoclosure` is _unnecessary_ for the wrapped value itself. That is, the wrapped-value argument at the call-site will _always_ be wrapped in a call to `init(wrappedValue:)`, which can already support `@autoclosure` arguments. Furthermore, using `@autoclosure` for the backing wrapper-type would be misleading, since the type spelled out for the parameter would not be the true autoclosure type; consider the `reportProgress` example from above with an `@autoclosure` parameter:
+
+```swift
+func postUrl(
+  @Lowercased urlString: @autoclosure () -> String
+) { ... }
+
+
+postUrl(urlString: "mySite.xyz/myUnformattedUsErNAme")
+```
+The above code would be transformed to:
+```swift
+func postUrl(
+  urlString _urlString: @autoclosure () -> Lowercased
+) { ... }
+
+
+postUrl(urlString: Lowercased(wrappedValue: "mySite.xyz/myUnformattedUsErNAme"))
+```
+As you can see, the changing type of the `@autoclosure` is incredibly misleading, as it's not obvious that `@autoclosure` applies to the backing wrapper rather than the wrapped value. For these reasons, using `@autoclosure` for a property-wrapper function-parameter will be invalid.
```

Comment:
@hborla, could you clarify what you mean by "the changing type of the `@autoclosure`". Do you mean the type that results from adding the `@autoclosure` attribute, the action of adding the attribute, or something else entirely?